### PR TITLE
macOS gamepad support

### PR DESCRIPTION
**Pull request by kotleni**
## Description
Implemented gamepad support on macOS. This works with my  <a href="https://github.com/kotleni/VirtualHID-macOS">driver for virtual HID devices</a> (you can copy this repository for yourself, the code is not licensed). Vibration not supported. I also added the IOKit dependency (shipped with Xcode).

_Sunshine host tested on Macbook Air M1 (2020).
Tested on two clients (Moonlight-nx (Nintendo Switch), and Moonlight (iOS))._

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

File: src/platform/macos/input.cpp [hunk 1]
```diff
@@ -282,16 +366,99 @@ void unicode(input_t &input, char *utf8, int size) {
...
+  if(flags & X) gamepad.buttons |= GAMEPAD_BTN_X;
+  if(flags & Y) gamepad.buttons |= GAMEPAD_BTN_Y;
+  if(flags & DPAD_UP) gamepad.buttons |= GAMEPAD_BTN_UP;
+  if(flags & DPAD_DOWN) gamepad.buttons |= GAMEPAD_BTN_DOWN;
+  if(flags & DPAD_LEFT) gamepad.buttons |= GAMEPAD_BTN_LEFT;
+  if(flags & DPAD_RIGHT) gamepad.buttons |= GAMEPAD_BTN_RIGHT;
+
+  // Map triggers
+  if(gamepad_state.lt > 0) gamepad.buttons |= GAMEPAD_BTN_LT;
+  if(gamepad_state.rt > 0) gamepad.buttons |= GAMEPAD_BTN_RT;
```
Comment (cgutman) [inline src/platform/macos/input.cpp:450]: Is it possible to emulate these with analog axes? Digital triggers are problematic for driving games and other games where precise trigger control is required.

File: src/platform/macos/input.cpp [hunk 2]
```diff
@@ -282,16 +366,99 @@ void unicode(input_t &input, char *utf8, int size) {
...
+
+  IOObjectRelease(virtgamepad_iterator);
+
+  if(!found) {
+    BOOST_LOG(info) << "Gamepad: Unable to open IOService."sv;
+    return 1;
+  }
+
+  // Fill up the input arguments.
+  virtgamepad_input[0] = (uint64_t)strdup(VIRTGAMEPAD_NAME);   // device name
```
Comment (cgutman) [inline src/platform/macos/input.cpp:397]: It looks like your driver is identifying the gamepads by name. Can we concatenate the `nr` parameter to the hardcoded name or something to support multiple gamepads?

**Review by cgutman — COMMENTED** [review]
Kexts require some special signing entitlements, right?

File: src/platform/macos/input.cpp (hunk 2 above)
Comment (kotleni) [reply src/platform/macos/input.cpp:397]: This name can be changed, it's dynamic.

File: src/platform/macos/input.cpp (hunk 1 above)
Comment (kotleni) [reply src/platform/macos/input.cpp:450]: I'm not sure that now I can implement analog support.

**Comment by kotleni** [comment]
> Kexts require some special signing entitlements, right?

It's needed only for building. I can make public release on GitHub.

File: src/platform/macos/input.cpp (hunk 1 above)
Comment (kotleni) [reply src/platform/macos/input.cpp:450]: And, I don't have device with analog trigger, lol.

**Comment by kotleni** [comment]
I implemented analog triggers (digital works too), but it needs to be tested.

_P.S. Sorry for the wrong name of the last commit._

File: src/platform/macos/input.cpp (hunk 1 above)
Comment (kotleni) [reply src/platform/macos/input.cpp]: Updated. See comments in PR.

**Comment by kotleni** [comment]
It didn't work. I will make another PR in the future with support for analog triggers.

**Comment by kotleni** [comment]
What about merging?

**Comment by ReenigneArcher** [comment]
Thank you for this PR! It might take me a few days to get this merged.

In the meantime can you update the documentation?

- remove statements that gamepad doesn't work on macOS
- add statements to usage and troubleshooting (similar to windows) about how to use gamepads on macOS

In your other repo, please improve the installation instructions for someone who would install the release version.

I'd like to have a few macOS users test this as well before it's merged.

**Comment by kotleni** [comment]
@ReenigneArcher 
**I did it.** 
Due to the complexity of installing kernel extensions, this will be a problem for normal users.

**Comment by kotleni** [comment]
@ReenigneArcher 
**I did it.** 
Due to the complexity of installing kernel extensions, this will be a problem for normal users.

**Comment by brad-richardson** [comment]
What kind of game compatibility do you expect from this? When I looked at adding controller support previously it seemed like games either had to 1) explicitly look for your controller (so needed to emulate something popular like 360/ps4 controllers) or 2) use GCController (which we can’t provide a source input for). I know very little about this space but would like to learn.

Awesome work though, I’ll take it for a spin this weekend!

**Comment by kotleni** [comment]
@brad-richardson 
I guess macOS games use <a href="https://developer.apple.com/documentation/gamecontroller">GameController</a> to implement gamepad support. If so, then there will be no problems with support.
All games launched through Steam work fine. I've tested a couple more games/apps outside of Steam, they work great too.
Regarding the emulation of popular gamepads, this is really possible. Perhaps in the future (if necessary) we will do this.

It needs testing by other users.

File: docs/source/about/installation.rst [hunk 3]
```diff
@@ -152,7 +152,7 @@ Uninstall:
 
 macOS
 -----
-Sunshine on macOS is experimental. Gamepads do not work. Other features may not work as expected.
+Sunshine on macOS is experimental. Other features may not work as expected.
```
Comment (ReenigneArcher) [inline docs/source/about/installation.rst:155]: Let's just remove this whole line.

**Review by ReenigneArcher — COMMENTED** [review]
Can you also add:

`notes-append "For gamepad support, install VirtualHID from 'https://github.com/kotleni/VirtualHID-macOS/releases/latest'`

to the Portfile, just before the final note with the documentation link.

**Comment by kotleni** [comment]
@ReenigneArcher 
Done.

**Comment by brad-richardson** [comment]
I tried this out and wasn't able to get it working. I only see "Gamepad: Unable to open IOService" when I run sunshine.

I took a couple of additional steps that I suspect are unique to M1 macs or macOS 13:
1. It only showed a prompt to load the extension if I ran the installer script while the settings app was already opened to privacy & security
2. When I attempted to load the extension (with SIP already disabled) it said I had to enable an additional setting in recovery mode. This was at Utilities -> Startup Security -> "Allow user management of kernel extensions"

Even after "enabling" the extension and restarting as prompted it seems to not have taken effect - eg `kextstat | grep virthd` returns nothing. 

I think this is fine to ship as is for a first pass, especially if you think this could be a known problem for arm macs or the new OS. I would be interested in debugging further though

**Comment by kotleni** [comment]
> I tried this out and wasn't able to get it working. I only see "Gamepad: Unable to open IOService" when I run sunshine.
> 
> 
> 
> I took a couple of additional steps that I suspect are unique to M1 macs or macOS 13:
> 
> 1. It only showed a prompt to load the extension if I ran the installer script while the settings app was already opened to privacy & security
> 
> 2. When I attempted to load the extension (with SIP already disabled) it said I had to enable an additional setting in recovery mode. This was at Utilities -> Startup Security -> "Allow user management of kernel extensions"
> 
> 
> 
> Even after "enabling" the extension and restarting as prompted it seems to not have taken effect - eg `kextstat | grep virthd` returns nothing. 
> 
> 
> 
> I think this is fine to ship as is for a first pass, especially if you think this could be a known problem for arm macs or the new OS. I would be interested in debugging further though

I tested on the same configuration.

The second point is normal behavior, macOS says about it - that's why I did not write about it in the instructions. 

Have you tried re-running the installer after the system has restarted? Did the Apple logo flicker a couple of times on reboot?

File: packaging/macos/Portfile [hunk 4]
```diff
@@ -63,3 +63,4 @@ notes-append "Run @PROJECT_NAME@ by executing 'sunshine <path to user config>',
 notes-append "The config file will be created if it doesn't exist."
 notes-append "It is recommended to set a location for the apps file in the config."
 notes-append "See our documentation at 'https://docs.lizardbyte.dev/projects/sunshine/en/v@PROJECT_VERSION@/' for further info."
+notes-append "For gamepad support, install VirtualHID from 'https://github.com/kotleni/VirtualHID-macOS/releases/latest'
```
Comment (ReenigneArcher) [inline packaging/macos/Portfile:66]: please, move up one line (just above the documentation link) and leave a blank new line at the end of the file

File: docs/source/about/installation.rst [hunk 5]
```diff
@@ -152,8 +152,6 @@ Uninstall:
 
 macOS
 -----
-Sunshine on macOS is experimental. Other features may not work as expected.
-
```
Comment (ReenigneArcher) [inline docs/source/about/installation.rst:156]: leave one blank line between

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

**Comment by kotleni** [comment]
@ReenigneArcher Done.

File: docs/source/about/usage.rst [hunk 6]
```diff
@@ -140,6 +140,8 @@ Sunshine needs access to `uinput` to create mouse and gamepad events.
 
 macOS
 ^^^^^
+For gamepad support, install `VirtualHID <https://github.com/kotleni/VirtualHID-macOS/releases/latest>`_
+
 Sunshine can only access microphones on macOS due to system limitations. To stream system audio use
 `Soundflower <https://github.com/mattingalls/Soundflower>`_ or
 `BlackHole <https://github.com/ExistentialAudio/BlackHole>`_ and
```
Comment (ReenigneArcher) [inline docs/source/about/usage.rst:147]: I missed this before. Need to remove L152.

`.. Caution:: Gamepads are not currently supported.`

Comment (kotleni) [reply docs/source/about/usage.rst:147]: 🅰️

File: docs/source/troubleshooting/macos.rst [hunk 7]
```diff
@@ -11,3 +11,14 @@ If you get this error:
...
          launchctl load -w /Library/LaunchAgents/org.freedesktop.dbus-session.plist
+
+No gamepad detected
+-------------------
+#. Verify that you've installed `VirtualHID <https://github.com/kotleni/VirtualHID-macOS/releases/latest>`_.
+
+Unable to create HID device
+---------------------------
+If you get this error:
+   `Gamepad: Unable to create HID device. May be fine if created previously.`
```
Comment (ReenigneArcher) [inline docs/source/troubleshooting/macos.rst:22]: This doesn't format properly (https://lizardbyte--756.org.readthedocs.build/projects/sunshine/en/756/troubleshooting/macos.html#unable-to-create-hid-device). Try 4 spaces before this line.

Comment (kotleni) [reply docs/source/troubleshooting/macos.rst:22]: Check it, last commit

Comment (ReenigneArcher) [reply docs/source/troubleshooting/macos.rst:22]: I think you misunderstood my statement.

```rst
Unable to create HID device
---------------------------
If you get this error:
    `Gamepad: Unable to create HID device. May be fine if created previously.`
```

Comment (kotleni) [reply docs/source/troubleshooting/macos.rst]: Lol, sorry. Now, all okay?

Comment (ReenigneArcher) [reply docs/source/troubleshooting/macos.rst]: Looks good now! https://lizardbyte--756.org.readthedocs.build/projects/sunshine/en/756/troubleshooting/macos.html#unable-to-create-hid-device

Comment (kotleni) [reply docs/source/troubleshooting/macos.rst]: 🤗

**Comment by brad-richardson** [comment]
@kotleni Your release `installer.py` doesn't have a valid download URL on `GITHUB_RELEASE`, I assume you meant to add `VERSION` within the `{}`. It would be good to output the logs from the subprocess calls there to make it easier to debug. I followed the steps manually, here's my error log at the load step:
```
Executing: /usr/bin/kmutil load -p /Users/brad/Downloads/VirtualHID.kext
Error Domain=KMErrorDomain Code=30 "Validating extension failed: Kext it.kotleni.virthid v1.0.0d1 in executable kext bundle it.kotleni.virthid at /Users/brad/Downloads/VirtualHID.kext:

Filesystem error: Invalid permissions 664 on /Users/brad/Downloads/VirtualHID.kext/Contents/_CodeSignature/CodeResources: should not be group/other writable
Filesystem error: Invalid permissions 664 on /Users/brad/Downloads/VirtualHID.kext/Contents/MacOS/VirtualHID: should not be group/other writable
Filesystem error: Invalid permissions 664 on /Users/brad/Downloads/VirtualHID.kext/Contents/Info.plist: should not be group/other writable" UserInfo={NSLocalizedDescription=Validating extension failed: Kext it.kotleni.virthid v1.0.0d1 in executable kext bundle it.kotleni.virthid at /Users/brad/Downloads/VirtualHID.kext:

Filesystem error: Invalid permissions 664 on /Users/brad/Downloads/VirtualHID.kext/Contents/_CodeSignature/CodeResources: should not be group/other writable
Filesystem error: Invalid permissions 664 on /Users/brad/Downloads/VirtualHID.kext/Contents/MacOS/VirtualHID: should not be group/other writable
Filesystem error: Invalid permissions 664 on /Users/brad/Downloads/VirtualHID.kext/Contents/Info.plist: should not be group/other writable}
```

Is it not properly signed? Did I do something bad with the permissions?

**Comment by kotleni** [comment]
> Is it not properly signed? Did I do something bad with the permissions?

Hello. It's error looks like wrong permissions.
You `chown`ed kext folder recursively?
And you have disabled SIP in system?

**Comment by kotleni** [comment]
> Is it not properly signed? Did I do something bad with the permissions?

Hello. It's error looks like wrong permissions.
You `chown`ed kext folder recursively?
And you have disabled SIP in system?

**Comment by kotleni** [comment]
@brad-richardson 
Hello. It's error looks like wrong permissions.
You chowned kext folder recursively?
And you have disabled SIP in system?

**Comment by brad-richardson** [comment]
>You chowned kext folder recursively?
>And you have disabled SIP in system?

Screenshot 2023-01-17 at 6 50 06 PM

**Comment by kotleni** [comment]
@brad-richardson 

Try it: `sudo chmod -R 777 ./VirtualHID.kext`

**Comment by kotleni** [comment]
@brad-richardson 
And about installer.py
It's works for me. Link is valid, because version puted later in `url = GITHUB_RELEASE.format(VERSION)`

**Comment by brad-richardson** [comment]
Yeah my bad, I missed the format below. I fixed the file permission issues by just copying into `/tmp` first like you do in your script.

I'm able to load the extension then enable and reboot like it instructs, but the extension is still just not there after I login. Any other ideas?

**Comment by kotleni** [comment]
@brad-richardson Yes, I joined to server now, find my in off topic channel.

**Comment by kotleni** [comment]
@brad-richardson Try `kextstat | grep -v com.apple` for printing all enabled kexts. This output have `it.kotleni.virthid`?

**Comment by ReenigneArcher** [comment]
Marked as draft, can merge once https://github.com/kotleni/VirtualHID-macOS/issues/1 is resolved.

**Comment by briankendall** [comment]
I can add that switching away from using a kext and to DriverKit system extension would allow users of modern macOS to install the driver without needing to disable SIP, rebooting, or otherwise change the security settings of their mac. The only thing they'd need to do is authorize loading the system extension in the System Preferences / Settings app.

Unfortunately this is Apple, so of course there are developer-hostile restrictions to doing this. It requires a paid developer account on top of having to ask for special permission from Apple to use the entitlement that grants an app permission to load a dext. The dext also needs to be bundled with a specific app, even if it's intended to be used across multiple apps. But getting over these hurdles will make it much more accessible to normal users.

**Comment by classicalwow** [comment]
@kotleni hi,did the kext support x64 gamepad? i install the kext,but not work

**Comment by kotleni** [comment]
@classicalwow I was never able to do it, my implementation works every other time, I stopped trying to implement it.

**Comment by classicalwow** [comment]
image
@kotleni excuse me, I really want to use this feature,i install the kexts and build kotleni:macos-gamepad-support from source,but got wrong,help me,plz.

**Comment by kotleni** [comment]
@classicalwow Yes, this is main problem with my implementation.
Unresolved bug in kernel extension :<

**Comment by iMonZ** [comment]
Is there another option to implement gamepad support for sunshine?

**Comment by kotleni** [comment]
@iMonZ Idk. Maybe we can use Steam Input (but it's not documented). :o

**Comment by iMonZ** [comment]
@kotleni  Could it be that kext were deprecated by macOS?
Could we maybe use SystemExtensions to implement this?
Otherwise its still possible to use kext but you need to disable a setting in the BIOS. So we could at least make just a proof of concept

**Comment by kotleni** [comment]
> @kotleni  Could it be that kext were deprecated by macOS?
> Could we maybe use SystemExtensions to implement this?
> Otherwise its still possible to use kext but you need to disable a setting in the BIOS. So we could at least make just a proof of concept 

Hey. Idk what is "System extension", but my kext was very unstable.
macOS don't want to load it for some users, or it's can be loaded only once. And it's hard to debug this, bcs I can't see any logs. :)

So, maybe u need to recheck it? 🤔

**Comment by DeltriDev** [comment]
is this something that still exists? I would love to test it!

**Comment by ReenigneArcher** [comment]
No, but it might be easier to implement this now: https://developer.apple.com/documentation/gamecontroller/gcvirtualcontroller

**Comment by kayakyakr** [comment]
> No, but it might be easier to implement this now: https://developer.apple.com/documentation/gamecontroller/gcvirtualcontroller?language=objc

Seems to be targetted to creating a controller in iOS. Wouldn't necessarily be able to give us a controller on mac.

**Pull request closed without merging**